xcode/Subconscious/Shared/Services/OpenAiService.swift

```diff
@@ -0,0 +1,215 @@
+//
+//  OpenAiService.swift
+//  Subconscious (iOS)
+//
+//  Created by Ben Follington on 6/3/2024.
+//
+
+import Foundation
+import KeychainSwift
+
+enum OpenAIError: Error, Equatable {
+    case missingAPIKey
+    case invalidAPIKey
+    case invalidAPIUrl
+    case malformedRequestBody
+    case requestFailed(String)
+}
+
+extension OpenAIError {
+    public var localizedDescription: String {
+        switch self {
+        case .invalidAPIKey:
+            return "Invalid API key"
+        case .invalidAPIUrl:
+            return "Invalid API URL"
+        case .malformedRequestBody:
+            return "Malformed request body"
+        case .missingAPIKey:
+            return "Missing API key, please enter it in the settings"
+        case let .requestFailed(error):
+            return "Request failed: \(error)"
+        }
+    }
+}
+
+struct OpenAIKey: Equatable, Hashable {
+    @Redacted var key: String
+}
+
+struct LLMPrompt {
+    public let system: String
+    public let instruction: String
+}
+
+actor OpenAIService {
```

Comment:
A followup I'd like us to do:

- I'd like to think about how to "protocolize" this API surface to avoid getting two entangled with the specifics of one model.
- The chat API is a natural fit here. Most of the API vendors have converged on something nearly exactly the same shape as [OpenAI's chat API](https://platform.openai.com/docs/api-reference/chat/create#chat-create-response_format).
- The service should probably talk back and forth in a `Codable` `AIChat` type that can be serialized/deserialized to JSON. Similar shape to the chat API request/response, with an array of messages with `system`, `assistant`, and `user` roles, but free of model specifics.

Reply:
One more consideration. Let's think about how this might route through our `PromptService`. In principle, I like the idea of routing everything through it as a dispatcher.